src/ir/table-utils.h

```diff
@@ -73,6 +73,9 @@ inline Index append(Table& table, Name name, Module& wasm) {
   auto tableIndex = segment.data.size();
   segment.data.push_back(name);
   table.initial = table.initial + 1;
+  if (wasm.dylinkSection) {
+    wasm.dylinkSection->tableSize++;
```

Comment:
Could the tableSize instead be computed dynamically based on the size of segment[0].    Otherwise all table operations would need to mimic this behaviour.

Reply:
I don't think that would be valid in general. The memorySize and tableSize fields don't just indicate the size of a singleton segment, they also represent an allocation that the program wants. So the loader reserves that space for them. Also, there might not be a singleton segment.

But it's true that for emscripten output we can assume a singleton segment and also that there is no reservation of space in the table (unlike memory). So I guess this `append` method should be documented as emscripten-specific.

(Concretely, if we make binaryen write `tableSize = segment[0]` at all times during binary writing, that would mean it can only load dynamic libraries generated by emscripten.)

Reply:
For the memory size we do include extra space for BSS data.  For table size the is no analog of BSS and the tableSize in dyliib section is always the sizeof segment[0].    lld will never produce a binary with more then one table segment.    Maybe for other languages these features would make sense, but not for llvm output.

In any case we should at least assert that tableSize is dylink section is < total number of table elements?  That way is one accidentally adds table element that invlidates the dylink section  invariant we can fail.

Reply:
If you have a module with multiple table segments and you hit this code things are probably going break.   Imagine two segments, one following the other.  If you grow the first then you start to clobber the second.   I think we should just mandate that dynamic libraries (at least those the binaryen can parse) have a single table segment that is equal in size of the tableSize declared in the dylink section.

Reply:
Yeah, we should assert at least, good point. I made it a fatal error now, and added comments explaining the issue more.